[100% review] Document .taskcluster.yml version 1

djmitche opened the pull request:
Thoughts on this documentation?

As written, I feel like it's kind of dumping users in the deep end: write a YAML file that takes this obscure Github JSON object and finagles it into this obscure Taskcluster Task JSON object.  Good luck!

But, I don't like what happened with the env vars, where they seemed to proliferate and weren't especially well-defined, and meant different things for different events.

Any thoughts on an elegant way to make users' lives easier without just creating an endless pit of utility functions?

djmitche committed:
Document .taskcluster.yml version 1

djmitche committed:
Document .taskcluster.yml version 1

imbstack commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,93 @@
+
+* `as_slugid` - a function which, given a label, will generate a slugid. 
+  Multiple calls with the same label for the same event will generate the same
+  slugiid, but different slugids in different events.  Use this to generate
```
slugiid

imbstack commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,93 @@
+# Policies
+
+The `policy` property defines policies for what is allowed on the repository.
+Policies are always read from the default branch (generally `master`) of the
```
Nice explanation

imbstack commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,93 @@
+Most projects prefer to run tasks for each pull request, so that the review
+process can take into account the task results. But if your project requires
+some secret data, or uses some expensive service, to test a pull request, then
+you probably do not want to run tasks for pull requests written by aritbrary
```
aritbrary

imbstack reviewed (dismissed):
I agree this feels like the deep end a bit. Honestly as a first cut it is probably helpful to give them lots of copy-pasta examples and maybe we can work on creating something easier from there?

imbstack reviewed (commented):
Oops, 10% review, so changes not approved! But the ideas so far are.

djmitche commented:
That's kind of what I was thinking -- let's get our feet wet with v1, and see what commonalities we come up with and add those later.  Then we have a nice firm "you can do anything" base.

owlishDeveloper committed:
Added examples, moved stuff to the new files

owlishDeveloper commented:
(I had to fuss with the branch a bit, because I forgot to rebase it against master)

I think it's better to get rid of the description of the `version 0` altogether, so I moved some things to the `version 1` file. I also added some examples, as I currently understand what the `version 1` should look like. 

The one thing I'm the most unsure about is environment variables. Why would we need utility functions for them?

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+
+# Policies
+
+The `policy` property defines policies for what is allowed on the repository. Policies are always read from the default branch (generally `master`) of the repository. This prevents a malicious contributor from changing the policy applied to a pull request in the pull request itself.
```
Will this involve a JSON-e rendering?  If so, we need to define what the context looks like for that rendering.  If not, then we need to say that the rendering only applies to the `tasks` property, and not the `policy` property.

I'm leaning toward not rendering right now, for simplicity - @imbstack what do you think?

imbstack replied:
Rendering that would be confusing/scary! I agree, let's just render what is in `tasks`.

owlishDeveloper replied:
I'll make sure that's clear in the doc. I'll also change the structure a bit and move the Policies section up, so that the rendering and tasks definition became neighbours.

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+
+ ### Github Events
+
+You can modify a task definition so that it will only run for specific Github events, by setting `task.extra.github.events` to a list of event names:
```
Instead of using `extra.github`, I think we should require users to do this themselves using JSON-e expressions.  `{$if: 'tasks_for == 'github-push' && event.head.branch == 'development', then: ..}`

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+  source: ${event.head.repo.url}
+tasks:
+  - provisionerId: ${taskcluster.docker.provisionerId}
+    workerType: ${taskcluster.docker.workerType}
```
I think we want to get away from supplying these values.  We have never changed them, and probably never will, so we might as well just tell our users what provisionerId / workerType to use.

imbstack replied:
Totally agreed.

owlishDeveloper replied:
Sounds cool!

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+      command:
+        - "/bin/bash"
+        - "-lc"
+        - "git clone ${event.head.repo.url} repo && cd repo && git checkout ${event.head.sha} && yarn global add node-gyp && yarn && yarn test"
```
++

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+  name: "TaskCluster GitHub Tests"
+  description: "All non-integration tests for taskcluster github"
+  owner: ${event.head.user.email}
+  source: ${event.head.repo.url}
```
There's no more top-level metadata :)

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+      owner: ${event.head.user.email}
+      source: ${event.head.repo.url}
+    extra:
+      github:
```
Let's try to get rid of this section..

djmitche reviewed (commented):
> I think it's better to get rid of the description of the version 0 altogether, so I moved some things to the version 1 file. 

I think it makes sense for v0 to be short, but we'll have a lot of people still using v0 files for a while, and they will want some way to figure out what those files mean.  

> The one thing I'm the most unsure about is environment variables. Why would we need utility functions for them?

It may be difficult or verbose to generate some of those values directly in JSON-e from the webhook payload.  For example, I think GITHUB_HEAD_TAG is generated by splitting the `refs/tags/` off the start of the head ref.

We may not end up needing these at all, or maybe only a few.  I think we should implement a v1 without them, or with only a few, and then try it out for some of the Taskcluster projects.  We will likely find some common annoyances -- and we can write utility functions for those.

imbstack commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,189 @@
+---
+
+Your main interface to Taskcluster-Github is via `.taskcluster.yml` in the root
+of your project. This is a YAML file that speciifies the tasks to run and when.
```
speciifies

imbstack reviewed (commented):
This is looking great.

owlishDeveloper committed:
Implemented review suggestions

Replaced extra.github section with if-else, removed supplying of provisionerId and workerType, removed top-level metadata, destroyed a typo

owlishDeveloper committed:
Reorganization

imbstack commented on docs/taskcluster-yml-v1.md:
````diff
@@ -0,0 +1,161 @@
+```yaml
+version: 1
+tasks:
+  - {$if: ' tasks_for == "github-push" ', then: {
````
I think maybe it would be nicer if this was written as

```
tasks:
  - $if: tasks_for == "github-push"
    then:
      ...
```

owlishDeveloper replied:
Oh, ok. I'll try that syntax

owlishDeveloper replied:
Yeah, that's much nicer! ❤️

imbstack reviewed (commented):
This looks great. Should we leave some more of the `version 0` bits around though since they won't change?

owlishDeveloper committed:
Nicer if-else syntax

owlishDeveloper committed:
Moar version 0

djmitche commented on docs/taskcluster-yml-v0.md:
```diff
@@ -3,14 +3,10 @@ title: .taskcluster.yml version 0
-**NOTE**: version-0 `.taskcluster.yml` files are deprecated. For any new
-projects, and to take full advantage of Taskcluster-Github features, use a
-[version 1](taskcluster-yml-v1) file.
+**NOTE**: version-0 `.taskcluster.yml` files are deprecated. For any new projects, and to take full advantage of Taskcluster-Github features, use a [version 1](taskcluster-yml-v1) file.
```
++

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,161 @@
+
+Policies are not rendered as a part of a task.
+
+##### Pull Requests
```
As a sub-heading of Policies, this can be `##`

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,161 @@
+
+# JSON-e Rendering
+
+The tasks defined in the YAML file are rendered using
```
Let's be really precise here: The `tasks` property of `.taskcluster.yml` is rendered..".  Then users will know that they can say something like:

```yaml
tasks:
  $let
    x: 10
  in:
    ...
```

that is, `tasks` need not be a literal array, as long as it JSON-e renders to an array.

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,161 @@
+# JSON-e Rendering
+
+The tasks defined in the YAML file are rendered using
+[JSON-e](https://github.com/taskcluster/json-e). You can view it as a `template`. The following `context` variables are provided:
```
Let's use `*` instead of backticks here, as emphasis: "You can view it as a *template*. The following *context* variables are provided:"

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,161 @@
+ 
+### Github Events
+
+You can put a task definition inside an `${if}, then` statement so that it will only run for specific Github events:
```
`{$if, then}` (brace before $)

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,161 @@
+
+* `as_slugid` - a function which, given a label, will generate a slugid. Multiple calls with the same label for the same event will generate the same slugid, but different slugids in different events.  Use this to generate taskIds, etc.
+
+## Task definition
```
Maybe "Examples"?

owlishDeveloper replied:
Do you mean rename this section or add moar examples?

djmitche replied:
Sorry, I mean rename the section.  That's lots of examples already :)

owlishDeveloper replied:
I like "Task definition" too... In the future we might want to add some task-related information in there. Besides, I have an "Example" section at the end. So I am proposing "Task definition and examples"

djmitche replied:
Sure, sounds good.

djmitche reviewed (commented):
This looks good!  Some minor formatting bits below.  You might also want to look at the outline of the document to see if the headings make sense.  One good way to do that is `cat docs/taskclutser-yml-v1.md | grep '^#'` to see just the headings and their levels.

Anyway, IMHO it's time to get implementin'!

owlishDeveloper committed:
Dustin's suggestions - first batch

owlishDeveloper committed:
Changes title

owlishDeveloper commented:
I started implementing, and the first question I have is: I have to validate the yaml according to the schema in the tests, don't I? My understanding is we're gonna have a different schema 🤔

imbstack commented:
@owlishDeveloper which schema in particular? We can always make a new schema!

djmitche committed:
mention that schedulerId is set by the tool

owlishDeveloper commented:
My commits are gone 🙀

djmitche commented:
Did you push something to this branch?  I pushed 89160ff within the last few weeks, but may have force-pushed and deleted something in the process.  Do you still have those changes locally?

owlishDeveloper commented:
Yes, I have them locally!

djmitche commented:
@owlishDeveloper when you get a chance, please push them to this branch.  I apologize for not seeing them earlier.  then we'll get this merged!

owlishDeveloper commented:
ok. Will tend to this at lunch!

owlishDeveloper commented:
There are merge conflicts, so will be sorting out in the evening.

owlishDeveloper committed:
Merge Dustin's branch

owlishDeveloper committed:
mention that schedulerId is set by the tool

owlishDeveloper commented:
Done! I think it needs one final review tho

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,180 @@
+
+* `as_slugid` - a function which, given a label, will generate a slugid. Multiple calls with the same label for the same event will generate the same slugid, but different slugids in different events.  Use this to generate taskIds, etc.
+
+## Task Definition and Examples
```
A short section on tags might be useful here, since it's not clear in the github documentation.

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,180 @@
+### Github Events
+
+You can put a task definition inside an `$if` - `then` statement so that it will only run for specific Github events:
+
```
We should probably note here that your template should be careful to produce nothing for unrecognized events (a caution we failed to follow in Firefox's .taskcluster.yml)

djmitche commented on docs/taskcluster-yml-v1.md:
````diff
@@ -0,0 +1,180 @@
+```yaml
+version: 1
+tasks:
+  - $if: ' event.head.branch == "master" '
````
this should be `tasks_for == "github-push" && event.head.branch == "master"`

djmitche commented on docs/taskcluster-yml-v1.md:
```diff
@@ -0,0 +1,180 @@
+tasks:
+  - $if: ' tasks_for == "github-push" || tasks_for == "github-pull-request"  && event.head.branch != "master" '
+    then:
+      provisionerId: 'aws-provisioner-v1'
```
This task, and the one below, need to specify a taskId :)

djmitche reviewed (commented):
I'll make these updates shortly (and not force-push over your changes this time!)

djmitche committed:
minor docs updates

owlishDeveloper commented on docs/taskcluster-yml-v1.md:
```diff
@@ -74,20 +76,42 @@ tasks:
+tasks:
+  - $if: 'tasks_for == "github-push"'
+    then:
+      $if: 'event.ref[:10] == "refs/tags/"'
```
what does `[:10]` mean here?

djmitche replied:
It's a slice, and it takes the prefix of the ref to compare.  The whole expression is equivalent to `event.ref.startswith('refs/tags/')`, but there's no `startswith` operator in JSON-e.

owlishDeveloper replied:
wow, that's cool!

imbstack reviewed (approved):
Sweet. We probably shouldn't land this till we land the little fixes in the try-v1 branch, but otherwise looks great!
